### PR TITLE
Ak/i210 master merge

### PR DESCRIPTION
**Pull request by AboudyKreidieh**
## Pull request information

- **Status**: ? (ready to merge / in development)
- **Kind of changes**: ? (bug fix / new feature / documentation...)
- **Related PR or issue**: ? (optional)

## Description

? (general description)

### CODE REVIEW

File: flow/controllers/car_following_models.py [hunk 1]
```diff
@@ -56,15 +56,18 @@ def __init__(self,
                  v_des=8,
                  time_delay=0.0,
                  noise=0,
-                 fail_safe=None):
+                 fail_safe=None,
+                 display_warnings=True):
```
Comment (eugenevinitsky) [inline flow/controllers/car_following_models.py:60]: Do these display a crazy amount of warnings? If so we might want to tset it to False

File: flow/core/kernel/simulation/traci.py [hunk 2]
```diff
@@ -184,3 +269,66 @@ def teardown_sumo(self):
             os.killpg(self.sumo_proc.pid, signal.SIGTERM)
         except Exception as e:
             print("Error during teardown: {}".format(e))
+
+    def save_emission(self, run_id=0):
+        """Save any collected emission data to a csv file.
+
+        If not data was collected, nothing happens. Moreover, any internally
```
Comment (eugenevinitsky) [inline flow/core/kernel/simulation/traci.py:276]: nit: if  no data

File: flow/core/kernel/simulation/traci.py [hunk 3]
```diff
@@ -184,3 +269,66 @@ def teardown_sumo(self):
...
+        stored data by this class is clear whenever data is stored.
```
Comment (eugenevinitsky) [inline flow/core/kernel/simulation/traci.py:277]: nit: is cleared

File: flow/core/kernel/vehicle/base.py [hunk 4]
```diff
@@ -139,21 +139,8 @@ def apply_acceleration(self, veh_id, acc):
...
-        smoothing.
-
-        Parameters
-        ----------
-        veh_id : str or list of str
-            list of vehicle identifiers
-        acc : float or array_like
-            requested accelerations from the vehicles
+        smooth : bool
+            whether to apply acceleration smoothly or not, default: True
```
Comment (eugenevinitsky) [inline flow/core/kernel/vehicle/base.py:143]: nit: this needs to be expanded upon. What it means for an accel to be smooth is not obvious

File: flow/core/kernel/vehicle/traci.py [hunk 5]
```diff
@@ -293,10 +293,8 @@ def _add_departed(self, veh_id, veh_type):
         self.__vehicles[veh_id]["type"] = veh_type
 
         # specify energy model
-        energy_model = \
-            self.type_parameters[veh_type]["energy_model"]
-        self.__vehicles[veh_id]["energy_model"] = \
-            energy_model[0](veh_id, **energy_model[1])
+        self.__vehicles[veh_id]["energy_model"] = self.type_parameters[
+            veh_type]["energy_model"]()
```
Comment (eugenevinitsky) [inline flow/core/kernel/vehicle/traci.py:297]: is this guaranteed to not be none?

File: flow/core/kernel/vehicle/traci.py [hunk 6]
```diff
@@ -1158,52 +1159,34 @@ def set_max_speed(self, veh_id, max_speed):
...
-        if "accel_with_noise_no_failsafe" not in self.__vehicles[veh_id]:
-            self.__vehicles[veh_id]["accel_with_noise_no_failsafe"] = None
-        return self.__vehicles[veh_id]["accel_with_noise_no_failsafe"]
+        metric_name = 'accel'
+        if noise:
+            metric_name += '_with_noise'
+        else:
+            metric_name += '_no_noise'
+        if failsafe:
+            metric_name += '_with_falsafe'
```
Comment (eugenevinitsky) [inline flow/core/kernel/vehicle/traci.py:1185]: nit: '_with_failsafe'

File: flow/core/rewards.py [hunk 7]
```diff
@@ -306,6 +306,61 @@ def punish_rl_lane_changes(env, penalty=1):
     return total_lane_change_penalty
 
 
+def energy_consumption(env, gain=.001):
+    """Calculate power consumption of a vehicle.
```
Comment (eugenevinitsky) [inline flow/core/rewards.py:310]: this calculates it for all vehicles

File: flow/core/rewards.py [hunk 8]
```diff
@@ -330,16 +385,16 @@ def instantaneous_mpg(env, veh_ids=None, gain=.001):
...
-            if gallons_per_hr > 0 and speed >= 0.0:
+            if speed >= 0.0:
                 cumulative_gallons += gallons_per_hr
                 cumulative_distance += speed
 
     cumulative_gallons /= 3600.0
     cumulative_distance /= 1609.0
     # miles / gallon is (distance_dot * \delta t) / (gallons_dot * \delta t)
-    mpg = cumulative_distance / cumulative_gallons
+    mpg = cumulative_distance / (cumulative_gallons + 1e-6)
```
Comment (eugenevinitsky) [inline flow/core/rewards.py:398]: This would still blow up at really small values, no?

File: flow/data_pipeline/data_pipeline.py [hunk 9]
```diff
@@ -10,37 +10,23 @@
...
-    Returns
-    -------
-    output_file_path : str
-        the local path of the outputted csv file
+    emission_files : list
+        paths to the SUMO emission
+    trajectory_table_path : str
+        path to the file fot S3 upload only
+    source_id : str
+        the source id of this simulation
```
Comment (eugenevinitsky) [inline flow/data_pipeline/data_pipeline.py:23]: this comment is not sufficient to clarify what this is

File: flow/data_pipeline/data_pipeline.py [hunk 10]
```diff
@@ -10,37 +10,23 @@
...
-        the name of the partition to put this output to
-
-    Returns
-    -------
-    output_file_path : str
-        the local path of the outputted csv file
+    emission_files : list
+        paths to the SUMO emission
+    trajectory_table_path : str
+        path to the file fot S3 upload only
```
Comment (eugenevinitsky) [inline flow/data_pipeline/data_pipeline.py:21]: nit: for s3 upload

File: flow/data_pipeline/data_pipeline.py [hunk 11]
```diff
@@ -10,37 +10,23 @@
...
-    # raw_output['partition'] = partition_name
-    raw_output = raw_output.sort_values(by=["time", "id"])
-    output_file_path = data_path[:-4]+"_trajectory.csv"
-    raw_output.to_csv(output_file_path, index=False)
-    return output_file_path
+    for i in range(len(emission_files)):
+        emission_output = pd.read_csv(emission_files[i])
+        emission_output['source_id'] = source_id
+        emission_output['run_id'] = "run_{}".format(i)
+        emission_output.to_csv(trajectory_table_path, mode='a+', index=False, header=(i == 0))
```
Comment (eugenevinitsky) [inline flow/data_pipeline/data_pipeline.py:29]: these conditions have some magic numbers and therefore need a comment

File: flow/envs/multiagent/base.py [hunk 12]
```diff
@@ -322,7 +322,3 @@ def apply_rl_actions(self, rl_actions=None):
         # clip according to the action space requirements
         clipped_actions = self.clip_actions(rl_actions)
         self._apply_rl_actions(clipped_actions)
-
-    def set_iteration_num(self):
-        """Increment the number of training iterations."""
-        self.num_training_iters += 1
```
Comment (eugenevinitsky) [inline flow/envs/multiagent/base.py:328]: kind of need this function for curricula. I suppose I can just move it into the subclasses

File: flow/envs/multiagent/i210.py [hunk 13]
```diff
@@ -244,7 +231,7 @@ def compute_reward(self, rl_actions, **kwargs):
             rewards = {rl_id: reward for rl_id in valid_ids}
 
         # curriculum over time-gaps
-        if self.headway_curriculum and self.num_training_iters <= self.headway_curriculum_iters:
+        if self.headway_curriculum and self._num_training_iters <= self.headway_curriculum_iters:
```
Comment (eugenevinitsky) [inline flow/envs/multiagent/i210.py:234]: thank you for doing this

File: flow/core/kernel/vehicle/traci.py (hunk 5 above)
Comment (AboudyKreidieh) [reply flow/core/kernel/vehicle/traci.py:297]: ya, it defaults to some energy model (it's never None)

File: flow/core/rewards.py (hunk 8 above)
Comment (AboudyKreidieh) [reply flow/core/rewards.py:398]: this is just a stabilizing term for when a simulation starts. we shouldn't have this problem after the warmup steps are passed

File: flow/envs/multiagent/base.py (hunk 12 above)
Comment (AboudyKreidieh) [reply flow/envs/multiagent/base.py:328]: this was moved to flow/envs/base.py. It can be accessed by multiagent environments as well.

**Pull request merged**